File: app/src/main/java/com/bignerdranch/stockwatcher/model/service/repository/StockDataRepository.java
package com.bignerdranch.stockwatcher.model.service.repository;

import com.bignerdranch.stockwatcher.model.service.StockInfoForSymbol;
import com.bignerdranch.stockwatcher.model.service.StockInfoResponse;
import com.bignerdranch.stockwatcher.model.service.StockService;
import com.bignerdranch.stockwatcher.model.service.StockSymbol;

import java.util.concurrent.TimeUnit;

import io.reactivex.Observable;
import timber.log.Timber;

public class StockDataRepository extends BaseRepository {

    private static final String CACHE_PREFIX_GET_STOCK_INFO = "stockInfo";
    private static final String CACHE_PREFIX_GET_STOCK_INFO_FOR_SYMBOL = "getStockInfoForSymbol";
    private static final String CACHE_PREFIX_GET_STOCK_SYMBOLS = "lookupStockSymbols";

    private final StockService service;

    public StockDataRepository(StockService service) {
        this.service = service;
    }

    public Observable<StockInfoForSymbol> getStockInfoForSymbol(String symbol) {
        Timber.i("method: %s, symbol: %s", CACHE_PREFIX_GET_STOCK_INFO_FOR_SYMBOL, symbol);
        Observable<StockInfoForSymbol> stockInfoForSymbolObservable = Observable.combineLatest(
                lookupStockSymbol(symbol),
                fetchStockInfoFromSymbol(symbol),
                StockInfoForSymbol::new);

        return cacheObservable(CACHE_PREFIX_GET_STOCK_INFO_FOR_SYMBOL + symbol, stockInfoForSymbolObservable);
    }

    //stock info request, which depends on the first result from lookup stock request
    private Observable<StockInfoResponse> fetchStockInfoFromSymbol(String symbol) {
        return lookupStockSymbol(symbol)
                .map(StockSymbol::getSymbol)
                .flatMap(this::getStockInfo);
    }

    private Observable<StockSymbol> lookupStockSymbol(String symbol) {
        Timber.i("%s, symbol: %s", CACHE_PREFIX_GET_STOCK_SYMBOLS, symbol);

        return cacheObservable(CACHE_PREFIX_GET_STOCK_SYMBOLS + symbol, service.lookupStock(symbol).cache());
    }

    private Observable<StockInfoResponse> getStockInfo(String symbol) {
        Timber.i("method: %s, symbol: %s", CACHE_PREFIX_GET_STOCK_INFO, symbol);
        Observable<StockInfoResponse> observableToCache = service
                .stockInfo(symbol).delay(3, TimeUnit.SECONDS).cache();

        return cacheObservable(CACHE_PREFIX_GET_STOCK_INFO + symbol, observableToCache);
    }

}

File: app/src/main/java/com/bignerdranch/stockwatcher/model/service/repository/StockDataRepository.kt
package com.bignerdranch.stockwatcher.model.service.repository

import com.bignerdranch.stockwatcher.model.service.StockInfoForSymbol
import com.bignerdranch.stockwatcher.model.service.StockInfoResponse
import com.bignerdranch.stockwatcher.model.service.StockService
import com.bignerdranch.stockwatcher.model.service.StockSymbol
import io.reactivex.Observable
import io.reactivex.functions.BiFunction
import timber.log.Timber
import java.util.concurrent.TimeUnit

private val CACHE_PREFIX_GET_STOCK_INFO_FOR_SYMBOL = "getStockInfoForSymbol"
private val CACHE_PREFIX_GET_STOCK_SYMBOLS = "lookupStockSymbols"
private val CACHE_PREFIX_GET_STOCK_INFO = "stockInfo"

class StockDataRepository(private val service: StockService) : BaseRepository() {

    fun getStockInfoForSymbol(symbol: String): Observable<StockInfoForSymbol> {
        Timber.i("method: %s, symbol: %s", CACHE_PREFIX_GET_STOCK_INFO_FOR_SYMBOL, symbol)
        val combineLatest = Observable.combineLatest(lookupStockSymbol(symbol),
                fetchStockInfoFromSymbol(symbol),
                BiFunction(::StockInfoForSymbol))
        return cacheObservable(CACHE_PREFIX_GET_STOCK_INFO_FOR_SYMBOL + symbol, combineLatest)
    }

    private fun fetchStockInfoFromSymbol(symbol: String): Observable<StockInfoResponse> {
        return lookupStockSymbol(symbol)
                .flatMap { stockSymbol -> getStockInfo(stockSymbol.symbol) }
    }

    private fun lookupStockSymbol(symbol: String): Observable<StockSymbol> {
        return lookupStockSymbols(symbol).flatMap { Observable.fromIterable(it).take(1) }
    }

    private fun lookupStockSymbols(symbol: String): Observable<List<StockSymbol>> {
        Timber.i("%s, symbol: %s", CACHE_PREFIX_GET_STOCK_SYMBOLS, symbol)
        return cacheObservable(CACHE_PREFIX_GET_STOCK_SYMBOLS + symbol, service.lookupStock(symbol).cache())
    }

    private fun getStockInfo(symbol: String): Observable<StockInfoResponse> {
        Timber.i("method: %s, symbol: %s", CACHE_PREFIX_GET_STOCK_INFO, symbol)
        val observableToCache = service.stockInfo(symbol).delay(3, TimeUnit.SECONDS).cache()
        return cacheObservable(CACHE_PREFIX_GET_STOCK_INFO + symbol, observableToCache)
    }


}